rfcs/transparent_ascription.md

```diff
@@ -0,0 +1,339 @@
+
+Finally, transparent signatures are the visible result in types of *transparent
+ascription* expressions, a form of ascription where all type equalities are
+preserved, as in:
+
+```ocaml
+(* opaque *)
+module X = (struct type t = int type u = bool end : sig type t end)
+(* -> *) module X : sig type t end
+
+  (* transparent *)
+module X = (struct type t = int type u = bool end :> sig type t end)
+(* -> *) module X : sig type t = int end
+```
+
+It is already implemented in SML (with the reverse syntax!), although it does
+not have the same consequences on the type system, as SML does not have
+applicative functors. It is not present in OCaml, but can be obtained with a
+normal (opaque) ascription where all the type sharing has been made explicit in
+the signature (which is cumbersome in practice):
+
+```ocaml
+(* opaque, but with annotations *)
+module X = (struct type t = int type u = bool end : sig type t end with type t = int)
+(* -> *) module X : sig type t = int end
+```
+
+## 1.d Resources
+
+List of some relevant discussions (PRs and issues):
+
+* [Allow functor application and functor arguments as module aliases #10435](https://github.com/ocaml/ocaml/pull/10435)
+* [Transparent ascription of modules #10612](https://github.com/ocaml/ocaml/pull/10612)
+* [Use Mp_present for module aliases #10673](https://github.com/ocaml/ocaml/pull/10673)
+* [Disallow aliases of functor arguments in module types #11460](https://github.com/ocaml/ocaml/pull/11460)
+
+List of relevant papers (chronological order):
+
+* [Type-level module aliases: independent and equal](https://www.math.nagoya-u.ac.jp/~garrigue/papers/modalias.pdf)
+* [F-ing modules](https://dl.acm.org/doi/abs/10.1145/1708016.1708028)
+* [Fulfilling OCaml Modules with Transparency](https://dl.acm.org/doi/abs/10.1145/3649818)
+* [Avoiding Signature Avoidance in ML Modules with Zippers](https://dl.acm.org/doi/10.1145/3704902)
+
+# 2. High-level summary of the change
+
+As of today, the aliasing mechanism merges several features in an unsatisfactory
+way. Splitting the features into user-controlled, well-behaved and
+well-understood ones will improve usability and expressivity, and simplify the
+typechecker. We propose the following changes:
+
+1. **Feature split**:
+   * Split aliases into *static aliases* (absent ones) and *dynamic aliases*
+     (present ones) in the internal representation of the typechecker.
+   * Provide attributes (`static_alias`, `dynamic_alias`), as well as a separate
+     syntax for each construction (syntax transition and backward compatibility
+     are detailed below).
+   * Do not change yet the "aliasable" criterion, i.e. keep the restriction that
+     dynamic aliases cannot contain functor parameter, functor application, etc.
+   * Remove the `presence` flag (which would coincide exactly with being a
+     static alias).
+   * Have dynamic aliases be a subtype of static aliases (a dynamic alias can be
+     coerced into a static one).
```

Comment:
From a distance I found this subtyping relation surprising -- aren't dynamic aliases allowed in more situations than static aliases?

Reply:
Maybe you could say "a dynamic alias can be coerced into a static one whenever it is also well-formed as a static alias".